---
params:
  title: "EPIC Array Quality-Control"
  author_name: "Firstname Lastname"
  author_affiliation: "Institution"
  author_email: "some@email.com"
  output_directory: "/EPIC_QC"
  csv_file: "sample_sheet_clean.csv"
  data_directory: "/idats"
  array: "EPIC"
  annotation: "ilm10b4.hg19"
  filter_snps: TRUE
  filter_non_cpg: TRUE
  filter_xy: TRUE
  filter_multihit: TRUE
  filter_beads: TRUE
  population: NULL
  bead_cutoff: 0.05
  detection_pvalues: 0.01
  filter_callrate: TRUE
  callrate_samples: 0.99
  callrate_probes: 1
  sex_threshold: NULL 
  sex_colname: NULL
  norm_background: "oob"
  norm_dye: "RELIC"
  norm_quantile: "quantile1" 
  cell_tissue: NULL
  pca: TRUE
  pca_vars: !r c("Sample_Plate", "Sentrix_ID")
title: '`r params[["title"]]`'
author:
- name: '`r params[["author_name"]]`'
  affiliation: '`r params[["author_affiliation"]]`'
  email: '`r params[["author_email"]]`'
date: '`r format(Sys.time(), "%B %d, %Y")`'
output:
  bookdown::html_document2:
    theme: simplex
    toc: true
    toc_depth: 3
    toc_float: 
      collapsed: false
    fig_width: 6.3
    fig_height: 4.7
    number_sections: true
    self_contained: true
editor_options: 
  chunk_output_type: console
---


```{r setup, include = FALSE}
options(stringsAsFactors = FALSE)


### Load packages ==================================================================================
suppressPackageStartupMessages({
  library("AnnotationHub")
  library("bookdown")
  library("dmapaq")
  library("doParallel")
  library("ENmix")
  library("ExperimentHub")
  library("FlowSorted.Blood.EPIC")
  library("FlowSorted.CordBloodCombined.450k")
  library("fs")
  library("ggplot2")
  library("ggrepel")
  library("ggtext")
  library("gt")
  library("knitr")
  library("minfi")
  library("parallel")
  library("rain")
  library("readr")
  library("RefFreeEWAS")
  library("rmarkdown")
  library("scales")
  library("stats")
  library("tibble")
  library("tidyr")
  library("utils")
})


### Functions ======================================================================================
`%>%` <- gt::`%>%`


### Environment ====================================================================================
output_directory <- params[["output_directory"]]
dir.create(output_directory, showWarnings = FALSE, recursive = TRUE, mode = "0777")


### Define theme ===================================================================================
ggplot2::theme_set(
  ggplot2::theme_light(base_size = 11) +
    ggplot2::theme(
      plot.title = ggtext::element_markdown(), 
      plot.title.position = "plot", 
      plot.subtitle = ggtext::element_markdown(face = "italic", size = ggplot2::rel(0.8)),
      plot.caption = ggtext::element_markdown(face = "italic", size = ggplot2::rel(0.5)), 
      plot.caption.position = "plot",
      axis.text.x = ggtext::element_markdown(), 
      axis.text.x.top = ggtext::element_markdown(), 
      axis.text.y = ggtext::element_markdown(), 
      legend.box.just = "left", 
      legend.key.size = ggplot2::unit(0.5, "lines"), 
      legend.text = ggtext::element_markdown(size = ggplot2::rel(0.5))
    )
)


### knitr settings =================================================================================
knitr::opts_chunk$set(
  results = "asis",
  include = TRUE,
  echo = FALSE,
  warning = FALSE,
  message = FALSE,
  dpi = 120,
  tidy = FALSE,
  crop = TRUE,
  autodep = TRUE,
  fig.align = "center"
)


### Sex check ======================================================================================
do_check_sex <- !is.null(params[["sex_colname"]])


### Cell heterogeneity =============================================================================
if (nchar(system.file(package = "FlowSorted.Blood.EPIC")) == 0) {
  do_cell_composition <- FALSE
  message('Package "FlowSorted.Blood.EPIC" is not available.')
} else {
  do_cell_composition <- !is.null(params[["cell_tissue"]])
}
```

# Methods and Parameters

* Array: `r params[["array"]]`
* Annotation package from bioconductor: `r params[["annotation"]]`

## Call Rate

* `filter_callrate`: `r params[["filter_callrate"]]`
* The threshold for the detection pvalues: `r params[["detection_pvalues"]]`
* The call rate threshold for samples: `r params[["callrate_samples"]]`  
    => should samples with less than the specified call rate (`r params[["callrate_samples"]]`) for detection p-values below $\alpha=`r params[["detection_pvalues"]]`$ be removed?
* The call rate threshold for probes: `r params[["callrate_probes"]]`  
    => should probes with less than the specified call rate (`r params[["callrate_probes"]]`) for detection p-values below $\alpha=`r params[["detection_pvalues"]]`$ be removed?

## Pre-Processing

* The method to estimate background normal distribution parameters: `r params[["norm_background"]]`  
    => method from [*ENmix*](https://doi.org/doi:10.18129/B9.bioc.ENmix)
* The dye bias correction: `r params[["norm_dye"]]`  
    => method from [*ENmix*](https://doi.org/doi:10.18129/B9.bioc.ENmix)
* The quantile normalisation: `r params[["norm_quantile"]]`  
    => method from [*ENmix*](https://doi.org/doi:10.18129/B9.bioc.ENmix)

## Probes Filtering

* `filter_snps`: `r params[["filter_snps"]]`  
    => should probes in which the probed CpG falls near a SNP be removed? ([Zhou et al., 2016](https://www.doi.org/10.1093/nar/gkw967))
* `filter_non_cpg`: `r params[["filter_non_cpg"]]`  
    => should non-cg probes be removed?
* `filter_xy`: `r params[["filter_xy"]]`  
    => should probes from X and Y chromosomes be removed?
* `filter_multihit`: `r params[["filter_multihit"]]` ([Nordlund et al., 2013](https://www.doi.org/10.1186/gb-2013-14-9-r105))  
    => should probes which align to multiple locations be removed?
* `filter_beads`: `r params[["filter_beads"]]`  
    => should probes with less than three beadcount in at least `r scales::percent(params[["bead_cutoff"]])` of the samples be removed?

## Sex Check

* The threshold value to discrimate sex: `r params[["sex_threshold"]]`  
    => flag samples with sex discrepancy.

## Ethnicity Check

* Name of the ethnicity population: `r if (is.null(params[["population"]])) "Not defined" else params[["population"]]`

## Cell Composition

* The cell tissue: `r params[["cell_tissue"]]`  
    => using a reference panel or `RefFreeCellMix` method from [*RefFreeEWAS*](https://cran.r-project.org/package=RefFreeEWAS)
    
## Final Processing

* Probe design type bias correction: `r "rcp"`  
    => method from [*ENmix*](https://doi.org/doi:10.18129/B9.bioc.ENmix)

* Batch effect normalisation: `r "ComBat"`  
    => method from [*ENmix*](https://doi.org/doi:10.18129/B9.bioc.ENmix)


# Quality Control

```{r sample_sheet}
sample_sheet <- readr::read_csv(params[["csv_file"]], skip = 8) %>% 
  dplyr::mutate(
    Sample_ID = Sample_Name,
    sex_clean = if (do_check_sex) {
      c("1" = 1, "2" = 2, "M" = 1, "F" = 2, "0" = 2)[ get(params[["sex_colname"]]) ]
    } else {
      NULL
    }
  ) %>% 
  dplyr::select(Sample_ID, dplyr::everything())

readr::write_csv(x = sample_sheet, path = file.path(tempdir(), "sample_sheet.csv"))

if (do_check_sex) {
  pca_vars <- intersect(colnames(sample_sheet), unique(c(params[["pca_vars"]], "sex_clean")))
} else {
  pca_vars <- intersect(colnames(sample_sheet), params[["pca_vars"]])
}
```

```{r read_idats}
output <- utils::capture.output({
  data_idats <- dmapaq::read_idats(
    directory = params[["data_directory"]], 
    csv_file = file.path(tempdir(), "sample_sheet.csv"), 
    meth_value_type = "B",
    filter_beads = params[["filter_beads"]], 
    bead_cutoff = params[["bead_cutoff"]], 
    filter_non_cpg = params[["filter_non_cpg"]],
    filter_snps = params[["filter_snps"]], 
    population = params[["population"]], 
    filter_multihit = params[["filter_multihit"]],
    filter_xy = params[["filter_xy"]], 
    detection_pvalues = params[["detection_pvalues"]],
    filter_callrate = params[["filter_callrate"]],
    callrate_samples = params[["callrate_samples"]],
    callrate_probes = params[["callrate_probes"]],
    norm_background = params[["norm_background"]],
    norm_dye = params[["norm_dye"]],
    norm_quantile = params[["norm_quantile"]],
    array_name = params[["array"]], 
    annotation_version = params[["annotation"]],
    n_cores = params[["n_cores"]]
  )
}, type = "message")
output <- output[grep(pattern = "[dmapaq] Filtering IDAT files ...", output, fixed = TRUE):length(output)]
output <- output[!grepl(pattern = "^[=[]", x = output)]
cat(output, sep = "\n\n")

data_mset <- data_idats[["mset"]]
phenotypes <- as.data.frame(data_mset@metadata[["phenotypes"]])

data_rgset <- data_idats[["rgset"]]

readr::write_rds(x = data_idats, path = file.path(output_directory, paste0(params[["array"]], "_idats.rds")))
```


```{r call_rate}
data_call_rate_samples <- phenotypes %>% 
	dplyr::select(Sample_ID, call_rate) %>% 
	dplyr::arrange(call_rate) %>% 
	dplyr::mutate(
		x = seq_along(call_rate),
		label = ifelse(call_rate < params[["callrate_samples"]], Sample_ID, NA)
	)
```
  
```{r call_rate_samples_tab}
phenotypes %>% 
  dplyr::mutate("< Call Rate Threshold" = list(c(0.90, 0.95, 0.97, 0.98, 0.99, 1))) %>% 
  tidyr::unnest("< Call Rate Threshold") %>% 
  dplyr::group_by(`< Call Rate Threshold`) %>% 
  dplyr::summarise(
    "Samples (N)" = sum(call_rate < `< Call Rate Threshold`),
    "Samples (%)" = signif(sum(call_rate < `< Call Rate Threshold`) / dplyr::n() * 100, digits = 3)
  ) %>%
  dplyr::arrange(dplyr::desc(`< Call Rate Threshold`)) %>% 
  gt::gt(auto_align = "center") %>% 
  gt::opt_row_striping() %>% 
  gt::tab_style(
    style = gt::cell_fill(color = "dodgerblue", alpha = 0.5), 
    locations = gt::cells_body(
      columns = gt::vars(`< Call Rate Threshold`, `Samples (N)`, `Samples (%)`), 
      rows = `< Call Rate Threshold` == params[["callrate_samples"]]
    )
  )
```

```{r call_rate_samples_fig}
ggplot2::ggplot(
  data = data_call_rate_samples,
  mapping = ggplot2::aes(x = x, y = call_rate, label = label)
) +
  ggplot2::geom_point(
    colour = scales::viridis_pal(begin = 0.5, end = 0.5)(1), 
    shape = 1,
    na.rm = TRUE
  ) +
  ggplot2::geom_hline(
    mapping = ggplot2::aes(yintercept = params[["callrate_samples"]]),
    colour = "black", 
    linetype = 2,
    na.rm = TRUE
  ) +
  ggrepel::geom_label_repel(
    min.segment.length = ggplot2::unit(0, "lines"), 
    size = 2, 
    segment.colour = "black",
    colour = "black",
    na.rm = TRUE
  ) +
  ggplot2::labs(x = "Samples", y = "Call Rate") +
  ggplot2::scale_y_continuous(label = scales::percent) +
	ggplot2::scale_x_continuous(label = scales::comma)
```

## Cell composition

```{r do_cell_composition, eval = do_cell_composition}
if (
  grepl("^blood$", params[["cell_tissue"]], ignore.case = TRUE) & 
    nchar(system.file(package = paste0("FlowSorted.Blood.", params[["array"]]))) == 0
) {
  do_cell_composition <- FALSE
  message(paste0('Package "', paste0("FlowSorted.Blood.", params[["array"]]), '" is not available.'))
}

if (
  grepl("^cordblood$", params[["cell_tissue"]], ignore.case = TRUE) & 
    nchar(system.file(package = "FlowSorted.CordBloodCombined.450k")) == 0
) {
  message('Package "FlowSorted.CordBloodCombined.450k" is not available.')
  if (
    nchar(system.file(package = "AnnotationHub")) == 0 &
    nchar(system.file(package = "ExperimentHub")) == 0
  ) {
    message('Experiment "FlowSorted.CordBloodCombined.450k" will be downloaded with "ExperimentHub".')
  } else {
    do_cell_composition <- FALSE
    message('Package "FlowSorted.CordBloodCombined.450k", "AnnotationHub" and "ExperimentHub" are not available.')
  }
}

if (
  (
    !grepl("^blood$", params[["cell_tissue"]], ignore.case = TRUE) &
      !grepl("^cordblood$", params[["cell_tissue"]], ignore.case = TRUE)
  ) &
    nchar(system.file(package = "RefFreeEWAS")) == 0
) {
  do_cell_composition <- FALSE
  message('Package "RefFreeEWAS" is not available.')
}
```

```{r do_cell_composition_msg}
if (!do_cell_composition) {
  cat("No cell tissue was provided or no available reference set (packages).\n")
} else {
  cat("IDOL optimised CpGs are used when available in the relevant reference set or method.\n")
}
```

```{r cell_composition, eval = do_cell_composition, results = "hide"}
cell_comp <- switch(
  EXPR = tolower(params[["cell_tissue"]]),
  "blood" = {
    idol_cpgs <- switch(params[["array"]],
      "450k" = FlowSorted.Blood.EPIC::IDOLOptimizedCpGs450klegacy,
      "EPIC" = FlowSorted.Blood.EPIC::IDOLOptimizedCpGs
    )
    out <- FlowSorted.Blood.EPIC::estimateCellCounts2(
      rgSet = data_rgset, 
      compositeCellType = "Blood",
      processMethod = "preprocessNoob", 
      probeSelect = "IDOL", 
      cellTypes = c("CD8T", "CD4T", "NK", "Bcell", "Mono", "Neu"),
      referencePlatform = paste0("IlluminaHumanMethylation", params[["array"]]),
      IDOLOptimizedCpGs = idol_cpgs, 
      returnAll = FALSE,
      meanPlot = FALSE, 
      verbose = FALSE
    )$counts
    colnames(out) <- paste0("CellT_", colnames(out))
    out
  }, 
  "cordbloodlegacy" = {
    out <- FlowSorted.Blood.EPIC::estimateCellCounts2(
      rgSet = data_rgset, 
      compositeCellType = "CordBlood",
      processMethod = "preprocessNoob", 
      probeSelect = "auto", 
      cellTypes = c("Bcell", "CD4T", "CD8T", "Gran", "Mono", "NK", "nRBC"),
      referencePlatform = "IlluminaHumanMethylation450k",
      IDOLOptimizedCpGs = NULL, 
      returnAll = FALSE,
      meanPlot = FALSE, 
      verbose = FALSE
    )$counts
    colnames(out) <- paste0("CellT_", colnames(out))
    out
  }, 
  "cordblood" = {
    if (nchar(system.file(package = "FlowSorted.CordBloodCombined.450k")) == 0) {
      idol_cpgs <- get(utils::data("IDOLOptimizedCpGsCordBlood", package = "FlowSorted.CordBloodCombined.450k"))
    } else {
      idol_cpgs <- NULL
    }
    exp_hub <- ExperimentHub::ExperimentHub()
    FlowSorted.CordBloodCombined.450k <- AnnotationHub::query(
      x = exp_hub, 
      pattern = "FlowSorted.CordBloodCombined.450k"
    )[[1]]
    out <- FlowSorted.Blood.EPIC::estimateCellCounts2(
      rgSet = data_rgset,
      compositeCellType = "cordbloodCombined",
      processMethod = "preprocessNoob",
      probeSelect = if (!is.null(idol_cpgs)) "IDOL" else "auto",
      cellTypes = c("Bcell", "CD4T", "CD8T", "Gran", "Mono", "NK", "nRBC"),
      referencePlatform = paste0("IlluminaHumanMethylation", params[["array"]]),
      referenceset = "FlowSorted.CordBloodCombined.450k",
      IDOLOptimizedCpGs = idol_cpgs, 
      returnAll = FALSE,
      meanPlot = FALSE, 
      verbose = FALSE
    )$counts
    colnames(out) <- paste0("CellT_", colnames(out))
    AnnotationHub::removeCache(exp_hub, ask = FALSE)
    out
  },
  {
    estimate_k_cluster <- function(Rmat, max_k = 25, n_cores = 1) {
      cluster <- parallel::makeCluster(n_cores)
      doParallel::registerDoParallel(cluster)

      svdRmat <- RefFreeEWAS::svdSafe(Rmat)
      tmp <- do.call("rbind", lapply(
        X = 0:max_k,
        mc_Rmat = Rmat,
        mc_svdRmat = svdRmat,
        FUN = function(Ktest, mc_Rmat, mc_svdRmat) {
          N1 <- dim(mc_Rmat)[1]
          N2 <- dim(mc_Rmat)[2]
          if (Ktest == 0) {
            tmpRminLU <- mc_Rmat
          } else {
            tmpRminLU <- mc_Rmat - mc_svdRmat$u[, 1:Ktest] %*% (mc_svdRmat$d[1:Ktest] * t(mc_svdRmat$v[, 1:Ktest]))
          }
          tmpSigSq <- rowSums(tmpRminLU * tmpRminLU) / N2
    
          c(
            K = Ktest,
            AIC = 2 * (N1 + Ktest * (N1 + N2)) + 
              N1 * N2 +
              N2 * sum(log(tmpSigSq)),
            BIC = log(N2) * (N1 + Ktest * (N1 + N2)) + 
              N1 * N2 + 
              N2 * sum(log(tmpSigSq))
          )
      }))

      parallel::stopCluster(cluster)

      list(
        icTable = tmp, 
        best = tmp[c(AIC = which.min(tmp[, "AIC"]), BIC = which.min(tmp[, "BIC"])), "K"],
        custom_best = tmp[c(
          AIC = which.max(abs(diff(tmp[, "AIC"])[-1])) + 1, 
          BIC = which.max(abs(diff(tmp[, "BIC"])[-1])) + 1
        ), "K"]
      )
    }
    beta_matrix <- stats::na.exclude(minfi::getBeta(data_mset))
    max_k <- min(ncol(beta_matrix), 25)
    k_estimated <- min(estimate_k_cluster(
      Rmat = beta_matrix, 
      max_k = max_k, 
      n_cores = min(params[["n_cores"]], max_k)
    )$best)
    mu0 <- RefFreeEWAS::RefFreeCellMixInitialize(
      Y = beta_matrix,
      K = k_estimated,
      Y.Distance = NULL, 
      Y.Cluster = NULL, 
      largeOK = TRUE,
      dist.method = "euclidean"
    )
    
    RefFreeCellMixObj <- RefFreeEWAS::RefFreeCellMix(
      Y = beta_matrix, 
      mu0 = mu0, 
      K = NULL,
      iters = 10, 
      Yfinal = NULL, 
      verbose = FALSE
    )

    out <- RefFreeCellMixObj[["Omega"]]
    colnames(out) <- paste0("CellT_", 1:ncol(out))
    out
  }
)

phenotypes <- dplyr::left_join(
  x = dplyr::mutate(phenotypes, Sample_ID = as.character(Sample_ID)),
  y = tibble::rownames_to_column(as.data.frame(cell_comp), "Sample_ID"), 
  by = "Sample_ID"
)
```

```{r cell_composition_fig, eval = do_cell_composition}
dmapaq::ggheatmap(
  data = phenotypes %>% 
    dplyr::select(Sample_ID, dplyr::starts_with("CellT_")) %>% 
    tibble::column_to_rownames(var = "Sample_ID") %>% 
    tidyr::drop_na(), 
  dendrogram_columns = TRUE,
  dendrogram_rows = TRUE,
  line_size = 0.5,
  font_size = params[["gg_fontsize"]],
  font_size_x = params[["gg_fontsize"]],
  font_size_y = 0, # params[["gg_fontsize"]] / 8,
  distance_method = "euclidean",
  cluster_method = "ward.D2",
  scale_name = "Cell Composition",
  scale_revert = FALSE
)
```

## Sex check

```{r do_check_sex}
if (!do_check_sex) {
  cat("No phenotypes for sex was provided.\n")
}
```

```{r sex_check_computation, eval = do_check_sex}
if (is.null(params[["sex_threshold"]])) {
  sex_predicted <- suppressWarnings(suppressPackageStartupMessages({
    minfi::getSex(minfi::mapToGenome(data_rgset), cutoff = -2)
  }))
  sex_density <- stats::density(sex_predicted$yMed - sex_predicted$xMed, n = 100000)
  
  min_diff_xy <- which(diff(sign(diff(sex_density$y))) == 2)
  min_diff_xy <- min_diff_xy[which.min(sex_density$y[min_diff_xy])]
  sex_threshold <- round(x = sex_density$x[min_diff_xy], digits = 3)
  
  # sex_threshold <- sex_density[c("x", "y")] %>% 
  #   as.data.frame() %>% 
  #   dplyr::mutate(local_minima = dplyr::lag(y) > y & dplyr::lead(y) > y) %>% 
  #   dplyr::filter(local_minima) %>% 
  #   .[["x"]]
} else {
  sex_threshold <- params[["sex_threshold"]]
}
sex_predicted <- suppressWarnings(suppressPackageStartupMessages({
  minfi::getSex(minfi::mapToGenome(data_rgset), cutoff = sex_threshold)
})) %>% 
  as.data.frame() %>% 
  dplyr::mutate(Sample_ID = minfi::sampleNames(data_rgset)) %>% 
  dplyr::rename(
    qc_xmedian = xMed,
    qc_ymedian = yMed,
    qc_predicted_sex = predictedSex
  )

cat("\nThe sex threshold used for prediction was:", sex_threshold, "\n\n")

phenotypes <- phenotypes %>% 
  dplyr::mutate(Sample_ID = as.character(Sample_ID)) %>% 
  dplyr::left_join(y = sex_predicted, by = "Sample_ID") %>% 
  dplyr::mutate(
    qc_predicted_sex = c("1" = 1, "2" = 2, "M" = 1, "F" = 2, "0" = 2)[qc_predicted_sex],
    qc_observed_sex = sex_clean,
    qc_sex_discrepancy = is.na(qc_observed_sex) | qc_observed_sex != qc_predicted_sex
  )
```

```{r sex_check_threshold, eval = do_check_sex}
ggplot2::ggplot(
  data = dplyr::mutate(phenotypes, diff = qc_ymedian - qc_xmedian),
  mapping = ggplot2::aes(x = diff)
) +
  ggplot2::geom_density(position = "identity", na.rm = TRUE) +
  ggplot2::geom_vline(xintercept = sex_threshold, linetype = 2, na.rm = TRUE) +
  ggplot2::labs(
    x = expression(atop(
      paste("Y chromosome median total intensity ", (log[2])),
      paste("- X chromosome median total intensity ", (log[2]))
    )),
    y = "Density"
  ) +
  ggplot2::scale_x_continuous(expand = ggplot2::expansion(mult = c(0, 0))) +
  ggplot2::scale_y_continuous(expand = ggplot2::expansion(mult = c(0, 0.05)))
```

```{r sex_check_fig, eval = do_check_sex}
annot_sex <- data.frame(
  qc_xmedian = c(c(0, 0, 20), c(0, 20, 20)),
  qc_ymedian = c(c(0, 20, 20), c(0, 0, 20)) + sex_threshold,
  qc_predicted_sex = rep(c(1, 2), each = 3)
)

axis_limits <- range(phenotypes[, c("qc_xmedian", "qc_ymedian")], na.rm = TRUE)

ggplot2::ggplot(
  data = phenotypes, 
  mapping = ggplot2::aes(
    x = qc_xmedian, 
    y = qc_ymedian, 
    shape = factor(qc_observed_sex), 
    colour = factor(qc_observed_sex)
  )
) +
  ggplot2::geom_polygon(
    mapping = ggplot2::aes(x = qc_xmedian, y = qc_ymedian, fill = factor(qc_predicted_sex)),
    alpha = 0.1,
    data = annot_sex,
    inherit.aes = FALSE
  ) +
  ggplot2::geom_abline(
    data = data.frame(Threshold = paste("=", sex_threshold), Seuil = sex_threshold), 
    mapping = ggplot2::aes(intercept = Seuil, slope = 1, linetype = Threshold),
    na.rm = TRUE
  ) +
  ggplot2::geom_point(
    data = ~ dplyr::filter(.x, !qc_sex_discrepancy), 
    size = 2, 
    na.rm = TRUE
  ) +
  ggplot2::geom_point(
    data = ~ dplyr::filter(.x, qc_sex_discrepancy),
    colour = "firebrick2",
    size = 4,
    show.legend = FALSE,
    na.rm = TRUE
  ) +
  ggrepel::geom_label_repel(
    data = ~ dplyr::filter(.x, qc_sex_discrepancy),
    mapping = ggplot2::aes(x = qc_xmedian, y = qc_ymedian, label = Sample_ID),
    segment.colour = "black",
    colour = "black",
    min.segment.length = ggplot2::unit(0, "lines"),
    size = 2,
    inherit.aes = FALSE,
    show.legend = FALSE,
    na.rm = TRUE
  ) +
  ggplot2::labs(
    x = expression(atop("X chromosome", paste("Median total intensity ", (log[2])))), 
    y = expression(atop("Y chromosome", paste("Median total intensity ", (log[2]))))
  ) +
  ggplot2::coord_cartesian(xlim = axis_limits, ylim = axis_limits) +
  ggplot2::scale_colour_viridis_d(
    name = "Sex (Predicted)", 
    breaks = c(1, 2), 
    labels = c("Male", "Female"),
    drop = FALSE,
    begin = 0.2, 
    end = 0.8
  ) +
  ggplot2::scale_fill_viridis_d(
    name = "Sex (Predicted)", 
    breaks = c(1, 2), 
    labels = c("Male", "Female"),
    drop = FALSE,
    begin = 0.2, 
    end = 0.8
  ) +
  ggplot2::scale_shape_manual(
    name = "Sex (Observed)", 
    breaks = c(1, 2), 
    labels = c("Male", "Female"), 
    values = c(22, 21),
    drop = FALSE
  ) +
  ggplot2::guides(
    fill = ggplot2::guide_legend(order = 1, override.aes = list(alpha = 1)), 
    shape = ggplot2::guide_legend(order = 2),
    linetype = ggplot2::guide_legend(order = 3),
    colour = "none"
  )
```

```{r discrepancy_table, eval = do_check_sex}
phenotypes %>%
  dplyr::count(qc_predicted_sex, qc_observed_sex) %>% 
  dplyr::mutate(
    qc_observed_sex = paste("Observed:", qc_observed_sex),
    qc_predicted_sex = paste("Predicted:", qc_predicted_sex)
  ) %>% 
  tidyr::spread(key = "qc_observed_sex", value = "n", fill = 0) %>% 
  dplyr::rename(` ` = qc_predicted_sex) %>% 
  gt::gt(auto_align = "center") %>% 
  gt::opt_row_striping()
```

```{r discrepancy_samples_table, eval = do_check_sex}
if (any(phenotypes[["qc_sex_discrepancy"]])) {
  phenotypes %>% 
    dplyr::filter(qc_sex_discrepancy) %>% 
    dplyr::select(Sample_Name, Sample_ID, qc_observed_sex, qc_predicted_sex) %>% 
    gt::gt(auto_align = "center") %>% 
    gt::opt_row_striping()
}
```

```{r normalisation, results = "hide"}
norm_beta <- ENmix::rcp(mdat = data_mset)
if (length(unique(minfi::pData(data_mset)[["Sentrix_ID"]])) > 1) { 
  norm_beta <- sva::ComBat(
    dat = norm_beta, 
    batch = factor(minfi::pData(data_mset)[["Sentrix_ID"]])
  )[rownames(data_mset), ]
}
colnames(norm_beta) <- minfi::pData(data_mset)[["Sample_ID"]]
if (min(norm_beta, na.rm = TRUE) <= 0) {
  norm_beta[norm_beta <= 0] <- min(norm_beta[norm_beta > 0])
}
if (max(norm_beta, na.rm = TRUE) >= 1) {
  norm_beta[norm_beta >= 1] <- max(norm_beta[norm_beta < 1])
}
data_mset@metadata[grep("_values", names(data_mset@metadata))] <- NULL
data_mset@metadata[["norm_beta_values"]] <- norm_beta
data_mset@metadata[["phenotypes"]] <- phenotypes
```

```{r export}
readr::write_rds(
  x = data_mset, 
  path = file.path(output_directory, paste0(params[["array"]], "_QC_mset.rds"))
)
readr::write_csv(
  x = tibble::rownames_to_column(as.data.frame(norm_beta), "cpg_id"), 
  path = gzfile(file.path(output_directory, paste0(params[["array"]], "_QC_betavalues.csv.gz")))
)
readr::write_csv(
  x = as.data.frame(phenotypes), 
  path = file.path(output_directory, paste0(params[["array"]], "_QC_phenotypes.csv"))
)
```

<!-- ## Principal Component Analysis -->
```{r pca, eval = params[["pca"]]}
cat("\n## Principal Component Analysis\n\n")
list_beta <- c(
	"raw_beta" = "Raw $beta$-values",
	"norm_beta" = "ComBat normalised $beta$-values"
)
data_batch <- list(
	"raw_beta" = `colnames<-`(minfi::getBeta(data_rgset), minfi::pData(data_rgset)[, "Sample_ID"]),
	"norm_beta" = data_mset@metadata[["norm_beta_values"]]
)

for (ibeta in seq_along(list_beta)) {
	cat("\n###", list_beta[ibeta], " {.tabset .tabset-fade .tabset-pills}\n\n")
	rain::pca_report(
		data = data_batch[[names(list_beta)[ibeta]]], 
		design = phenotypes, 
		id_var = "Sample_ID",
		technical_vars = pca_vars, 
		n_comp = min(10, ncol(data_batch[[names(list_beta)[ibeta]]])), 
		fig_n_comp = min(3, ncol(data_batch[[names(list_beta)[ibeta]]])),
		title_level = 4
	)
	cat("\n")
}
```

# R session information

```{r session_info, results = "markup"}
options("width" = 110)
sessioninfo::session_info()
```
